File: microservices/services/catalog-service/app/schemas/catalog.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime

class CategoryBase(BaseModel):
    name: str

class CategoryCreate(CategoryBase):
    pass

class CategoryOut(CategoryBase):
    id: int

    class Config:
        from_attributes = True

class LessonBase(BaseModel):
    title: str
    content_type: str = "TEXTO"
    content: Optional[str] = None
    sort_order: int = 1
    duration_minutes: int = 0

class LessonCreate(LessonBase):
    pass

class LessonOut(LessonBase):
    id: int
    module_id: int

    class Config:
        from_attributes = True

class ModuleBase(BaseModel):
    title: str
    sort_order: int = 1
    es_examen_modulo: bool = False
    es_examen_final: bool = False

class ModuleCreate(ModuleBase):
    pass

class ModuleOut(ModuleBase):
    id: int
    course_id: int

    class Config:
        from_attributes = True

class CourseBase(BaseModel):
    title: str
    description: str
    category_id: int
    price: float
    status: str = "PUBLISHED"
    nivel_dificultad: str = "PRINCIPIANTE"
    duracion_horas: int = 0
    es_certificacion: bool = False
    imagen_url: Optional[str] = None

class CourseCreate(CourseBase):
    pass

class CourseOut(CourseBase):
    id: int
    instructor_id: int

    class Config:
        from_attributes = True

class ModuleWithLessons(ModuleOut):
    lessons: List[LessonOut] = []

    class Config:
        from_attributes = True

class CourseDetailOut(CourseOut):
    modules: List[ModuleWithLessons] = []
    rating_count: int = 0
    average_rating: float = 0.0

    class Config:
        from_attributes = True

class RatingBase(BaseModel):
    score: int = Field(..., ge=1, le=5)
    comment: str = ""

class RatingCreate(RatingBase):
    pass

class RatingOut(RatingBase):
    id: int
    course_id: int
    user_id: int
    created_at: datetime

    class Config:
        from_attributes = True

class WishlistCreate(BaseModel):
    course_id: int

class WishlistItemOut(BaseModel):
    id: int
    user_id: int
    course_id: int
    added_at: datetime
    course: Optional[CourseOut] = None

    class Config:
        from_attributes = True

class LessonProgressOut(BaseModel):
    id: int
    user_id: int
    lesson_id: int
    completed: bool
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True

class CourseProgressOut(BaseModel):
    course_id: int
    total_lessons: int
    completed_lessons: int
    percentage: float
    completed: bool
    completed_lesson_ids: List[int] = []

class TrayectoriaBase(BaseModel):
    nombre: str
    descripcion: Optional[str] = None
    categoria_general: Optional[str] = None
    imagen_url: Optional[str] = None

class TrayectoriaCreate(TrayectoriaBase):
    pass

class TrayectoriaOut(TrayectoriaBase):
    id: int
    class Config:
        from_attributes = True

class TrayectoriaCursoOut(BaseModel):
    course_id: int
    sort_order: int
    course: Optional[CourseOut] = None
    class Config:
        from_attributes = True

class TrayectoriaDetailOut(TrayectoriaOut):
    cursos: List[TrayectoriaCursoOut] = []
    class Config:
        from_attributes = True

class TrayectoriaAddCurso(BaseModel):
    course_id: int
    sort_order: int = 1

# ── Announcements ─────────────────────────────────────────────────────────────

class AnnouncementBase(BaseModel):
    titulo: str
    descripcion: Optional[str] = None
    imagen_url: Optional[str] = None
    activo: bool = True

class AnnouncementCreate(AnnouncementBase):
    pass

class AnnouncementOut(AnnouncementBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/catalog.ts
export interface Category {
  id: number;
  name: string;
  description?: string;
}

export interface Lesson {
  id: number;
  title: string;
  video_url?: string;
  duration?: number;
  sort_order: number;
  content_type?: string;
}

export interface Module {
  id: number;
  title: string;
  sort_order: number;
  lessons: Lesson[];
  es_examen_modulo?: boolean;
  es_examen_final?: boolean;
}

export interface Course {
  id: number;
  title: string;
  description: string;
  category_id: number;
  price: number;
  status: string;
  nivel_dificultad?: string;
  duracion_horas?: number;
  instructor_id?: number;
  es_certificacion?: boolean;
}

export interface CourseDetail extends Course {
  modules: Module[];
  rating_count: number;
  average_rating: number;
}

export interface PaginatedResponse<T> {
  count: number;
  next: number | null;
  previous: number | null;
  results: T[];
}

export interface LessonProgress {
  id: number;
  user_id: number;
  lesson_id: number;
  completed: boolean;
  completed_at: string | null;
}

export interface CourseProgress {
  course_id: number;
  total_lessons: number;
  completed_lessons: number;
  percentage: number;
  completed: boolean;
  completed_lesson_ids: number[];
}

export interface TrayectoriaCurso {
  course_id: number;
  sort_order: number;
  course?: Course;
}

export interface Trayectoria {
  id: number;
  nombre: string;
  descripcion: string | null;
  categoria_general: string | null;
  cursos?: TrayectoriaCurso[];
}

// --- QUIZ & CERTIFICATION TYPES ---

export interface QuizOptionStudentOut {
  id: number;
  texto: string;
}

export interface QuizQuestionStudentOut {
  id: number;
  tipo: string;
  enunciado: string;
  puntaje: number;
  sort_order: number;
  options: QuizOptionStudentOut[];
}

export interface QuizDetailOut {
  id: number;
  lesson_id: number;
  titulo: string;
  puntaje_minimo_aprobacion: number;
  questions: QuizQuestionStudentOut[];
}

export interface QuizAttemptSubmitItem {
  question_id: number;
  respuesta_texto?: string;
  selected_option_id?: number;
}

export interface QuizAttemptSubmit {
  respuestas: QuizAttemptSubmitItem[];
}

export interface QuizAttemptAnswerOut {
  question_id: number;
  respuesta_texto?: string;
  selected_option_id?: number;
  similitud?: number;
  puntaje_obtenido: number;
  es_correcta?: boolean;
}

export interface QuizAttemptResultOut {
  id: number;
  quiz_id: number;
  user_id: number;
  puntaje_obtenido: number;
  puntaje_maximo: number;
  porcentaje: number;
  aprobado: boolean;
  created_at: string;
  answers: QuizAttemptAnswerOut[];
}

export interface CertificationModuleStatus {
  module_id: number;
  title: string;
  sort_order: number;
  es_examen_modulo: boolean;
  es_examen_final: boolean;
  bloqueado: boolean;
  aprobado: boolean | null;
}

export interface CertificationProgressOut {
  course_id: number;
  modules: CertificationModuleStatus[];
  certificado_disponible: boolean;
}
